refactor(semantic): make control flow generation optional.

rzvxa opened the pull request:
For maximum backward compatibility, we generate CFG by default.

Note: It can't be done with a simple method since lifetimes make it impossible(at least without unsafe trickery) I've tried to do it without a macro but it was just unintuitive.

rzvxa commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/3737?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#3737** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/3737?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#3728** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/3728?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rzvxa and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

Boshen commented on crates/oxc_semantic/src/builder.rs:
```diff
@@ -31,6 +31,16 @@ use crate::{
     Semantic,
 };
 
+macro_rules! cfg {
```
Rust analyzer breaks with macros, maybe `self.cfg.map(|cfg| {}`?

rzvxa replied:
With mapping, we can't have a single point for cfg use and it is much harder to disable it all together with flags. Maybe we can bikeshed it a bit?

rzvxa commented:
> Nice! Would be nice if we get rid of the macro.
> 
> I'll make it off by default in the next PR.

I'm working on making it disabled by default in the next few PRs, I have to first make linter tests to generate CFG optional(so we can mark it if we need cfg for the rule to work).

Boshen commented:
Ok, shall we merge or get rid of the macro?

rzvxa commented:
> Ok, shall we merge or get rid of the macro?

Do we want to feature gate the CFG in the future? If that's the case I think macro can help us, Otherwise, we should be fine with mapping the option.

I can also give the inline function another try, Maybe I can find a safe trick to make it work.

Boshen commented:
Ok, let's merge and iterate.

rzvxa committed:
refactor(semantic): make control flow generation optional. (#3737)

For maximum backward compatibility, we generate CFG by default.

Note: It can't be done with a simple method since lifetimes make it impossible(at least without unsafe trickery) I've tried to do it without a macro but it was just unintuitive.

overlookmotel commented on crates/oxc_semantic/src/builder.rs:
```diff
@@ -467,8 +484,11 @@ impl<'a> Visit<'a> for SemanticBuilder<'a> {
         program.scope_id.set(Some(self.current_scope_id));
 
         /* cfg */
```
If you can be bothered, I think all these `/* cfg */` comments could be removed, as the `cfg!` macro now performs same function of delimiting the code related to CFG. But this is just a style nit - feel free to ignore if you don't have time/don't agree.

rzvxa replied:
I thought of it but didn't do it since there are a handful of instances where we have cfg related code outside of this macro, Wanted to keep it as a convention for that reason, and I'd love to remove these when it is a bit more mature and we know all of our cfg logics can be contained in these macro scopes.

overlookmotel commented:
Oh poop. PR got merged while I was reviewing! @rzvxa I don't know if you think any of my comments above are worth considering and addressing in follow ups, or not?

rzvxa commented on crates/oxc_linter/src/rules/eslint/getter_return.rs:
```diff
@@ -55,16 +55,19 @@ declare_oxc_lint!(
 impl Rule for GetterReturn {
     fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        // control flow dependant
+        let Some(cfg) = ctx.semantic().cfg() else { return };
```
This was intentional, I was thinking that we can make the linter to omit CFG generation and disable the rules depending on it. Or the other way around if we have no rule dependent on CFG enabled we can disable the CFG generation altogether.
But we can go either way since it has no difference at the moment. I had no objection against unwrap/panics here.

overlookmotel replied:
> if we have no rule dependent on CFG enabled we can disable the CFG generation altogether.

That's a nice optimization. I think that way around is the better option, in which case it could be `unwrap` here.

I have a broader feeling that we need a convention for communicating in code when (1) something is actually optional vs (2) there's an invariant that means this branch *must* always go one way. When delving into parts of Oxc codebase which I'm new to, this ambiguity has left me scratching my head quite a few times, wondering "in what circumstances can this happen?", and it turns out the answer is "it can't".

`unwrap`/`unreachable!` is one way to communicate that, but it has the downside of bloating the ASM with panic code for the unreachable branch. So I don't have a good answer to this.

Anyway, it's a broader problem, so I think we can just leave it as is in this PR, since I was too late with my review.

rzvxa commented on crates/oxc_semantic/src/builder.rs:
```diff
@@ -31,6 +31,16 @@ use crate::{
     Semantic,
 };
 
+macro_rules! cfg {
```
Sure, I was thinking we can use std::cfg directly if needed, Named it `cfg` since it is a repeating pattern all over this file and it looks like the comments. I'll change it to `control_flow`.

overlookmotel replied:
Personally I find the use of "cfg" as abbreviation for "control flow graph" confusing in general. There's quite a strong convention in Rust that "cfg" is short for "config" e.g. `#[cfg(...)]`.

When I first started looking at semantic's code, I kept asking myself "what's this 'config' all about?". It didn't take me too long to figure it out, but I think it's useful to remember what your first encounter with the codebase felt like, to put yourself in the position of a new contributor (who of course we want to welcome). The principle of minimum surprise and all that...

That's a broader point, and not such a big deal. But on this specific one, personally I think overriding part of the standard prelude is a step too far! If you would be willing to change it, I think it'd be ideal.

rzvxa replied:
I've already changed it in a subsequent PR.

overlookmotel replied:
Ha! I am too slow again! Thanks for doing that.

rzvxa replied:
No worries mate! Winds are blowing fast, "Jibe Ho" ⚓

rzvxa commented:
@overlookmotel I've made some comments about the review conversations above, Let me know about your thoughts so I can integrate them in the next PR(s).
